utils/variant_qc.py

```diff
@@ -0,0 +1,346 @@
+        desc: bool = True
+) -> hl.Table:
+    """
+    Returns a table containing a binned rank for each row.
+    The bin is computed by dividing the `score_expr` into `n_bins` bins containing an equal number of elements.
+    This is done based on quantiles computed with hl.agg.approx_quantiles.
+    If a single value in `score_expr` spans more than one bin, the rows with this value are distributed
+    randomly across the bins it spans.
+
+    Notes
+    -----
+    The `rank_expr` defines which data the rank(s) should be computed on. E.g., to get an SNV rank and an Indel rank,
+    the following could be used:
+    rank_expr={
+       'snv_rank': hl.is_snp(ht.alleles[0], ht.alleles[1]),
+       'indels_rank': ~hl.is_snp(ht.alleles[0], ht.alleles[1])
+    }
+
+    :param Table ht: Input Table
+    :param NumericExpression score_expr: Expression containing the score
+    :param dict of str -> BooleanExpression rank_expr: Rank(s) to be computed (see notes)
+    :param int n_bins: Number of bins to bin the data into
+    :param int k: The `k` parameter of approx_quantiles
+    :param bool desc: Whether to bin the score in descending order
+    :return: Table with the binned ranks
+    :rtype: Table
+    """
+    import math
+
+    def quantiles_to_bin_boundaries(quantiles: List[int]) -> Tuple[DefaultDict, List[int], List[float]]:
+        """
+        Merges bins with the same boundaries into a unique bin while keeping track of
+        which bins have been merged and the global index of all bins.
+
+        :param quantiles: Original bins boundaries
+        :return: (dict of the indices of bins for which multiple bins were collapsed -> number of bins collapsed,
+                  Global indices of merged bins,
+                  Merged bins boundaries)
+        """
+
+        # Pad the quantiles to create boundaries for the first and last bins
+        bin_boundaries = [-math.inf] + quantiles + [math.inf]
+        bins_merged = defaultdict(int)
+
+        # If every quantile has a unique value, then bin boudaries are unique
+        # and can be passed to binary_search as-is
+        if len(quantiles) == len(set(quantiles)):
+            return (
+                bins_merged,
+                list(range(len(bin_boundaries))),
+                bin_boundaries
+            )
+
+        indexed_bins = list(enumerate(bin_boundaries))
+        i = 1
+        while i < len(indexed_bins):
+            if indexed_bins[i - 1][1] == indexed_bins[i][1]:
+                bins_merged[i - 1] += 1
+                indexed_bins.pop(i)
+            else:
+                i += 1
+
+        return (
+            bins_merged,
+            [x[0] for x in indexed_bins],
+            [x[1] for x in indexed_bins]
+        )
+
+    ht = ht.annotate(
+        **{f'_filter_{rid}': rexpr for rid, rexpr in rank_expr.items()},
+        _score=score_expr
+    )
+
+    logger.info(f'Adding rank using approximate_quantiles binned into {n_bins}, using k={10*n_bins}')
```

Comment:
So I'm not quite sure how `k` and `n_bins` play together to be honest.

Reply:
So I want to refrain setting them using a single parameter. The 1000 bins I've obtained from testing on the gnomAD data.

Reply:
Makes sense, I just didn't know if the logger should just write out what `k` was passed in rather than `10*n_bins`. Doesn't really make a difference for the code at all, just logs what k was actually passed int

Reply:
ohhhhh duh! Sorry, I should read my own code -- you're absolutely right!